### PR TITLE
Add ManualHasManyThrough preloader

### PR DESCRIPTION
**Pull request by xzhflying**

### CODE REVIEW

**Comment by xzhflying** [comment]
@lowjoel

File: lib/autoload/active_record/associations/preloader/manual_has_many_through.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+      @records_by_owner = records.each_with_object({}) do |record, h|
+        owner_id = record[source_reflection.foreign_key]
+        records = (h[owner_id] ||= [])
+        records << record
+      end
+    end
+
+    def associated_records_by_owner(preloader)
+      preloader.preload(owners,
```
Comment (lowjoel) [inline lib/autoload/active_record/associations/preloader/manual_has_many_through.rb]: Can you comment on what each line does?

Comment (lowjoel) [inline lib/autoload/active_record/associations/preloader/manual_has_many_through.rb]: Or better, break this up to get meaning. When we upgrade to Rails 5 I anticipate this needing to be fixed.

**Comment by lowjoel** [comment]
Specs?

File: lib/autoload/active_record/associations/preloader/manual_has_many_through.rb (hunk 1 above)
Comment (xzhflying) [inline lib/autoload/active_record/associations/preloader/manual_has_many_through.rb]: This method is copied from `HasManyThrough` itself. And I just modified one line. It makes sense to comment which line I modified but is that necessary to separate this method or comment each line?

File: lib/autoload/active_record/associations/preloader/manual_has_many_through.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        record_offset[record] = i
+      end
+
+      through_records.each_with_object({}) do |(lhs, center), records_by_owner|
+        pl_to_middle = center.group_by { |record| middle_to_pl[record] }
+
+        records_by_owner[lhs] = pl_to_middle.flat_map do |_pl, middles|
+          rhs_records = middles.flat_map do |r|
+            # For each source record, set the records which will be loaded
+            @records_by_owner[r.id]
```
Comment (xzhflying) [inline lib/autoload/active_record/associations/preloader/manual_has_many_through.rb]: https://github.com/rails/rails/blob/master/activerecord/lib/active_record/associations/preloader/through_association.rb#L50

**Pull request closed without merging**